### PR TITLE
9048 s3 storage

### PR DESCRIPTION
**Pull request by scttnlsn**

### CODE REVIEW

**Comment by scttnlsn** [comment]
Rebased against `develop` and fixed the merge conflict

File: app/controllers/media/objects_controller.rb [hunk 1]
```diff
@@ -1,79 +1,79 @@
...
-      type: @media_object.item_content_type,
-      disposition: disposition,
-      filename: media_filename)
-  end
+      authorize!(:show, @response) if @response
+
+      send_attachment(
+        @media_object.item,
+        style: style,
+        type: @media_object.item_content_type,
```
Comment (smoyth) [inline app/controllers/media/objects_controller.rb]: Don't think we are using this param anymore?

File: app/controllers/media/objects_controller.rb [hunk 2]
```diff
@@ -1,79 +1,79 @@
...
+        disposition: disposition,
+        filename: media_filename
+      )
```
Comment (smoyth) [inline app/controllers/media/objects_controller.rb]: Can you condense this into two lines?

File: app/controllers/operations_controller.rb [hunk 3]
```diff
@@ -22,7 +24,10 @@ def show
 
   def download
     if @operation.attachment.present?
-      send_file(@operation.attachment.path, filename: @operation.attachment_download_name)
+      send_attachment(
+        @operation.attachment,
+        filename: @operation.attachment_download_name
+      )
```
Comment (smoyth) [inline app/controllers/operations_controller.rb]: Condense please

File: app/controllers/questions_controller.rb [hunk 4]
```diff
@@ -78,10 +79,11 @@ def audio_prompt
 
     decorated_question = Odk::QuestionDecorator.decorate(@question)
 
-    send_file decorated_question.audio_prompt.path,
-      type: decorated_question.audio_prompt_content_type,
+    send_attachment(
+      decorated_question.audio_prompt,
       disposition: "attachment",
```
Comment (smoyth) [inline app/controllers/questions_controller.rb]: Maybe make disposition attachment the default to simplify things further? Inline is only rarely used I think?

File: app/helpers/answers_helper.rb [hunk 5]
```diff
@@ -15,16 +15,29 @@ def excerpt_if_exists(response, answer)
...
+  def media_path(object, params = {})
+    media_object_path(object, params.merge(type: object.kind.pluralize))
+  end
+
+  def thumb_path(object, params = {})
+    if object.dynamic_thumb?
+      media_path(object, params.merge(style: :thumb))
+    else
+      object.static_thumb_path
+    end
```
Comment (smoyth) [inline app/helpers/answers_helper.rb]: This turned out nice!

File: app/jobs/tabular_import_operation_job.rb [hunk 6]
```diff
@@ -19,4 +24,18 @@ def format_error_report(errors)
...
+    if file.options[:storage] == "fog"
+      # Preserve the file extension (Roo, for example, requires this in order to parse)
+      pathname = Pathname.new(file.path)
+      tmp = Tempfile.new(["tabular_import", pathname.basename.to_s])
+      tmp.binmode
+      URI.parse(file.expiring_url).open { |io| tmp.write(io.read) }
+      tmp.rewind
+      tmp
+    else
+      file
```
Comment (smoyth) [inline app/jobs/tabular_import_operation_job.rb]: Cool!

File: app/controllers/media/objects_controller.rb (hunk 1 above)
Comment (scttnlsn) [reply app/controllers/media/objects_controller.rb]: I added the following to the storage concern: `params[:type] ||= attachment.content_type` which is essentially the same as what's happening here.  I'll remove this line from the objects controller.

**Pull request merged**